test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp

```diff
@@ -0,0 +1,364 @@
+ * @param c Constraint that value must fulfill
+ * @param N Unused
+ */
+template <>
+void generateValues(double& v, Constraint c, int N) {
+  if (c == Constraint::None) {
+    v = -1;
+  } else if (c == Constraint::Positive) {
+    v = 1;
+  } else if (c == Constraint::NonNegative) {
+    v = 0;
+  }
+}
+
+/*
+ * Sets first element of v to a given value
+ *
+ * @tparam T1 type of v
+ * @param v Vector to assign values to
+ * @param value Value to assign
+ */
+template <typename T1>
+void setValue(T1& v, double value) {
+  v[0] = value;
+}
+
+/*
+ * Sets v equal to value
+ *
+ * @param v Variable to assign values to
+ * @param value Value to assign
+ */
+template<>
+void setValue(double& v, double value) {
+  v = value;
+}
+
+/*
+ * Return true if variables defined with constraint c1 do not violate constraint
+ * c2
+ *
+ * @param c1 First constraint
+ * @param c2 Second constraint
+ * @return True if values with constraint c1 and compatible with c2
+ */
+bool compatible(Constraint c1, Constraint c2) {
+  if (c1 == Constraint::NonNegative && c2 == Constraint::Positive)
+    return false;
+
+  if (c1 == Constraint::None && c2 == Constraint::NonNegative)
+    return false;
+
+  if (c1 == Constraint::None && c2 == Constraint::Positive)
+    return false;
+
+  return true;
+}
+
+/*
+ * twoParamCheckTrows feeds the given generate_samples callable various
+ * combinations of valid and invalid parameters (as defined by the two
+ * constraints, c1 and c2). For generated constraints that are not compatible
+ * with the supplied constraints, the generate_samples callable is expected
+ * to throw domain_errors.
+ *
+ * The generate_samples callable is also passed various other guaranteed invalid
+ * values like positive infinity, negative infinity, and NaNs (these should also
+ * cause domain_errors).
+ *
+ * The generated_samples callable is also tested to reject incompatibly sized
+ * input vectors. These should cause invalid_argument errors.
+ *
+ * The generate_samples callable must have the signature:
+ *   T_out generate_samples(T_param1 p1, T_param2 p2, T_rng)
+ * where T_param1 and T_param2 can be any combination of:
+ *   double, std::vector<double>, VectorXd, or RowVectorXd
+ * and T_rng is a boost random number generator
+ *
+ * The output of generate_samples must be of
+ *   stan::length(out) == stan::max_size(p1, p2)
+ *
+ * Each element of the output, out[i], should correspond to a random variate
+ * generated according to parameters p1[i] and p2[i]
+ *
+ * @tparam T_param1 Type of first parameter
+ * @tparam T_param2 Type of second parameter
+ * @tparam T_generate_samples Type of generate_samples functor
+ * @param generate_samples generate_samples functor to test
+ * @param c1 Constraint of first parameter
+ * @param c2 Constraint of second parameter
+ */
+template<typename T_param1, typename T_param2, typename T_generate_samples>
+void twoParamDistCheckThrows(T_generate_samples generate_samples, Constraint c1,
```

Comment:
There's nothing inherently distribution based about this function, is there? It might be acceptable as a framework for checking two parameter functions in general, which might suggest naming things more broadly and moving to a util directory? Also @bob-carpenter has some related test framework stuff, not sure if there's overlap here yet... maybe he can comment